fix(documentation): Update the README

chrislessard opened the pull request:
This patch makes a number of changes to the README, notably:

- Rephrasing of sentences and grammatical corrections
- Update of prerequisites
- General clarification of docker related steps
- Introduction of section on running the frontend

Resolves #42

freeatnet commented on docs/api/positions.yml:
```diff
@@ -0,0 +1,103 @@
+      tags:
+      - "positions"
+      summary: "Updates Position information for Position {id}"
+      parameters:
```
What attributes of a position can I update?

michellemtchai replied:
I actually tried adding the update-able parameters. I couldn't figure out how to add them without the Swagger Editor throwing errors.

freeatnet commented on docs/api/positions.yml:
```diff
@@ -0,0 +1,103 @@
+        description: "id for a Position in the Position Table"
+      responses:
+        404:
+          description: "Unable to update course data"
```
The 404 error appears to correspond to "A position with this ID does not exist".
We should also document what happens if the updated model does not pass validation.

freeatnet commented on docs/api/positions.yml:
```diff
@@ -0,0 +1,103 @@
+          schema:
+            type: "array"
+            items:
+              $ref: "#/definitions/Position"
```
👍

freeatnet commented on docs/api/positions.yml:
```diff
@@ -0,0 +1,103 @@
+      hours:
+        type: "integer"
+        example: 54
+      estimated_count:
```
We should add a description of this field.

michellemtchai replied:
Okay. I'll do that.

freeatnet commented on docs/api/positions.yml:
```diff
@@ -0,0 +1,103 @@
+      estimated_count:
+        type: "integer"
+        example: 2
+      estimated_total_hours:
```
Same with this.

michellemtchai replied:
Okay.

freeatnet commented on app/models/teach.rb:
```diff
@@ -0,0 +1,4 @@
+class Teach < ApplicationRecord
```
Do we plan on adding more fields in this model?

michellemtchai replied:
Probably not. Mostly, the information is going to come from either `Position` or `Instructor` model.

freeatnet replied:
In this case, we could take a look at the `has_and_belongs_to_many` relation between `Position` and `Instructor`

michellemtchai replied:
Wait. Does that mean that the 2 Models can actually be connected without the `Teach` model? Wow. Rails magic...

freeatnet replied:
http://guides.rubyonrails.org/association_basics.html#the-has-and-belongs-to-many-association

michellemtchai replied:
Okay. I'll do that.

michellemtchai replied:
I created a `instructors_positions` relation with a `has_and_belongs_to_many` on both the `Position` and `Instructor`. The table creation went well, but I can't seem to insert into the model. I used 
```
        instructor_ident = Instructor.find_by(name: instructor)
        if instructor_ident
          position.instructors << instructor_ident
```
`position` is a `Position` that was recently made. 

Basically,  it keeps saying `ActiveRecord::RecordInvalid: Validation failed: Position has already been taken`. I can't seem to get rid of this error. Suggestions?

freeatnet replied:
Hmm. Interesting. Before `position.instructors << instructor_ident`, could you log the values of `position.new_record?`, `position.valid?`, and `position.attributes`?

(for easy copy-pasting: `Rails.logger.debug "#{position.new_record?} #{position.valid?} #{position.attributes.inspect}"` :D)

michellemtchai replied:
``` Preference Load (0.2ms)  SELECT "preferences".* FROM "preferences" WHERE "preferences"."application_id" = $1 AND "preferences"."position_id" = $2  [["application_id", 22], ["position_id", 38]]
   (0.1ms)  SAVEPOINT active_record_1
  Position Load (0.1ms)  SELECT  "positions".* FROM "positions" WHERE "positions"."id" = $1 LIMIT $2  [["id", 38], ["LIMIT", 1]]
  SQL (0.3ms)  UPDATE "preferences" SET "rank" = $1, "updated_at" = $2 WHERE "preferences"."id" = $3  [["rank", 1], ["updated_at", "2017-06-21 19:23:35.982366"], ["id", 20]]
   (0.1ms)  RELEASE SAVEPOINT active_record_1
  Applicant Load (0.1ms)  SELECT  "applicants".* FROM "applicants" WHERE "applicants"."utorid" = $1 LIMIT $2  [["utorid", "applicant478"], ["LIMIT", 1]]
  Application Load (0.3ms)  SELECT  "applications".* FROM "applications" WHERE "applications"."applicant_id" = $1 LIMIT $2  [["applicant_id", 24], ["LIMIT", 1]]
  Position Load (0.1ms)  SELECT  "positions"."id" FROM "positions" WHERE "positions"."position" = $1 AND "positions"."round_id" = $2 LIMIT $3  [["position", "CSC100H1S"], ["round_id", 110], ["LIMIT", 1]]
  Preference Load (0.1ms)  SELECT  "preferences"."rank" FROM "preferences" WHERE "preferences"."application_id" = $1 LIMIT $2  [["application_id", 22], ["LIMIT", 1]]
   (0.1ms)  ROLLBACK
```
This is what I got.

freeatnet replied:
@michellemtchai Very odd! Could you push the code so I could take a run at it locally?

freeatnet replied:
When we're done removing this model, we can use `docker-compose run rails-app rails destroy model Teach` to clean the parts and pieces up :]

freeatnet commented on app/models/instructor.rb:
```diff
@@ -1,4 +1,5 @@
 class Instructor < ApplicationRecord
   has_many :courses
-  validates :email, uniqueness: true
```
Do we allow multiple instructors with the same email now? It seems that while we removed this line, the database uniqueness constraint [remains](https://github.com/uoft-tapp/tapp/pull/43/files#diff-1acd2e7e27a227829d5d14a91c863bb6R70).

michellemtchai replied:
The uniqueness was not allowing the `Instructor` to be inserted in the `seeds.rb` because apparently having `nil` for each instructor's email is not unique. This was a temporary fix because I didn't know if it would be a good idea to have real email listed on GitHub.

freeatnet replied:
Re: nil: having the email missing is a real possibility in production, we could use `uniqueness: {allow_nil: true}` here, and we should also make sure that we don't use an empty string as an email.

Re: real emails: since we already have names, we could generate fake emails as `"prof.#{name.gsub(/( )/, '.').downcase}@tapp-test.cs.toronto.edu"` or something ¯\_(ツ)_/¯

michellemtchai replied:
Okay. Though, I do actually have all of their email. I just don't know if it's a good idea to have them in GitHub.

freeatnet replied:
I agree, let's not put a full list of emails on open GitHub. We'll figure something out for production purposes.

michellemtchai replied:
Okay. Then, I'll just add the `uniqueness: {allow_nil: true}`. Should I just leave the `email` part alone for now?

freeatnet replied:
Sorry, that was meant to read `validates :email, uniqueness: {allow_nil: true}`. The option comes from the actual [validator method](http://api.rubyonrails.org/classes/ActiveRecord/Validations/ClassMethods.html#method-i-validates_uniqueness_of) underneath.

michellemtchai replied:
Okay.

freeatnet commented on db/migrate/20170601160156_create_positions.rb:
```diff
@@ -1,8 +1,12 @@
-      t.string :course_code, index: true, null: false
-      t.text :title
+      t.string :position, index: true, null: false
+      t.integer :round_id, index: true, null: false
```
Are multiple positions with the same title within the same `round_id` possible?

michellemtchai replied:
No. I'll fix this.

michellemtchai replied:
Though, actually, how do I do uniqueness for each `round_id`. Check constraints?

freeatnet replied:
We could use the [`scope` option](http://api.rubyonrails.org/classes/ActiveRecord/Validations/ClassMethods.html#method-i-validates_uniqueness_of) on the model side and a [two-column unique index](http://api.rubyonrails.org/classes/ActiveRecord/ConnectionAdapters/SchemaStatements.html#method-i-add_index-label-Creating+a+unique+index) on the migration/DB side.

michellemtchai replied:
Okay. Thanks.

freeatnet commented on app/services/chass_importer.rb:
```diff
@@ -167,6 +146,16 @@ def insert_positions
+
+      course_entry["instructors"].each do |instructor|
+        instructor_ident = Instructor.find_by(name: instructor)
+        if instructor_ident != nil
```
`nil` is always falsy.

michellemtchai replied:
What should I do instead? I didn't want the ChassImporter to simply stop in the middle of import. Doing `Instructor.find_by!(name: instructor)` only leads to a Rails error that jumps out of the importing. Is there an equivalent to `continue;` in Rails?

freeatnet replied:
Everything is good logic-wise, I'm just saying that `if instructor_ident != nil` is equivalent to `if instructor_ident` in this case.

michellemtchai replied:
Wow. More Rails magic... Thanks. I'll add that.

freeatnet replied:
@michellemtchai That's not really rails magic… `find_by` returns either a record or `nil` if it's not found. `nil` is a falsy value in Ruby (and most other languages), so we don't need to check if a value is not `nil` to establish it's truthiness.

freeatnet commented on app/services/chass_importer.rb:
```diff
@@ -167,6 +146,16 @@ def insert_positions
+        instructor_ident = Instructor.find_by(name: instructor)
+        if instructor_ident != nil
+          instructor_id = instructor_ident.id
+          teach ||= position.teaches.build(instructor_id: instructor_id)
```
We can use `instructor.id` directly here.

michellemtchai replied:
Okay. I'll do that.

freeatnet replied:
Also, is `teach ||=` necessary here?

michellemtchai replied:
Okay. I also didn't notice that. I'll remove the `teach ||=` part. Thanks.

freeatnet commented on app/services/chass_importer.rb:
```diff
@@ -66,99 +65,79 @@ def insert_application
+          position_row = Position.where(position_ident).select(:id).take
+          Rails.logger.debug position_row
+
+          if position_row != nil
```
Ditto on `nil`.

michellemtchai replied:
I don't understand what 'Ditto' means. Could you explain?

freeatnet replied:
> dit·to
> noun
> 1. used in accounts and lists to indicate that an item is repeated (often indicated by a ditto mark under the word or figure to be repeated).
> 2. a similar thing; a duplicate.
— Google :D

michellemtchai replied:
I still don't understand it within this context... Sorry.

freeatnet replied:
Hehe. I use ditto to point out multiple occurrences of the same issue. Here, that the `nil` is falsy, like in the [previous comment](https://github.com/uoft-tapp/tapp/pull/43#discussion_r123261435) (did GitHub rearrange them? 😒 ).

michellemtchai replied:
Yeah. GitHub did rearrange them. The falsy comment has been rearranged to like the second last comment while this one is somewhere near the top...

freeatnet replied:
Sorry, my bad, should've checked for that!

freeatnet commented on app/services/chass_importer.rb:
```diff
@@ -66,99 +65,79 @@ def insert_application
+    parse_preference(preferences).each do |preference|
+      position_ident = {position: preference, round_id: @round_id}
+      position = Position.where(position_ident).select(:id).take
+      if position != nil
```
Ditto on `nil`.

freeatnet commented on app/services/chass_importer.rb:
```diff
@@ -66,99 +65,79 @@ def insert_application
-      exists= "Course #{course_id} already exists"
-      insertion_helper(course, condition, exists)
-    end
+  def parse_preference(pref)
```
Have we agreed on the format here?

michellemtchai replied:
I just assumed it's going to be a CSV format. It wasn't specified on the site, so I just went with the format I made in the `mock_chass.json`. I can always update this once we do have a decided format.

freeatnet replied:
Alright, that works for me :+1:

freeatnet commented on app/services/chass_importer.rb:
```diff
@@ -66,99 +65,79 @@ def insert_application
-    end
+  def parse_preference(pref)
+    list = pref.split(',')
+    return list
```
We could rewrite this as
```
def parse_preference(pref)
  pref.split(',')
end
```

michellemtchai replied:
Okay. I'll do that.

freeatnet commented on app/services/chass_importer.rb:
```diff
@@ -66,99 +65,79 @@ def insert_application
+      if position != nil
+        position_id = position.id
+        preference_ident = {position_id: position_id}
+        pref ||= application.preferences.where(preference_ident).update(rank: 1)
```
Do we need to assign at all here?

michellemtchai replied:
Well, @reidka and I talked about it. And she said that there's going to be 2 levels of preference. The default rank is 2 and the preferred rank is 1. 

Actually, are you talking about the `position_id = position.id` part? I'll fix it.

freeatnet replied:
I'm just thinking the `pref ||=` part is unnecessary.

michellemtchai replied:
Oh. I didn't notice that. Thanks for catching it. I'll fix that.

freeatnet commented on db/migrate/20170601160156_create_positions.rb:
```diff
@@ -15,6 +15,7 @@ def change
       t.timestamps
     end
-
+    remove_index :positions, :position
```
If we don't want an index on `positions`, we should remove `index: true` from its column definition.

michellemtchai replied:
Okay. I'll add that.

freeatnet commented on db/migrate/20170601160156_create_positions.rb:
```diff
@@ -15,6 +15,7 @@ def change
     end
-
+    remove_index :positions, :position
+    add_index :positions, :position, unique: true
```
This doesn't consider that a position is unique within a round — and the same `position` value is valid when used with different `round_id`s.

michellemtchai replied:
Okay. I'll add that.

freeatnet replied:
Sorry, just to confirm, what are you planning to add here?

michellemtchai replied:
I plan to add ` add_index(:positions, [:position, :round_id], unique: true)`. Would it be right?

freeatnet replied:
@michellemtchai 💯

michellemtchai replied:
Okay.

Commit:
fix(documentation): Update the README

This commit makes a number of changes to the README, notably:

- Rephrasing of sentences and grammatical corrections
- Update of prerequisites
- General clarification of docker related steps
- Introduction of section on running the frontend

jmzaleski commented on README.md:
```diff
@@ -2,68 +2,117 @@
-To create and setup your local development database, simply navigate into the
-rails container and run the following rake task. It will create your local database,
-run all migrations, and populate the db with seed data. 
+To interact with the rails app, invoke `docker-compose exec tapp sh` and to
```
what means "interact" ??

chrislessard replied:
@jmzaleski It means that you can enter an interactive terminal session within the container, I took the wording from docker resources. Under such a session, you could run our rails specs, for example. Do you want this to be clarified further in the document ?

jmzaleski commented on README.md:
````diff
@@ -2,68 +2,117 @@
 $ psql tapp_development tapp
 ```
 
+## Running the frontend
````
I know this is a hard question.. but what is the "front end" in this usage?
I reckon it's how we serve the javascript of our app -- but I don't understand how we might start up the "front end" in production differently than from dev. 

Maybe I'm below the threshold of a reasonable knowledgeable reader?

chrislessard replied:
@jmzaleski Given how most people tend to use the term these days I don't think it will pose a problem. That said, the "frontend" here is how we serve the collection of *both* html, stylings (i.e. css) and javascript (note that rails does not serve any of these things). In other words, it's the presentational aspect of the project. 

I specified in the document that this is separate from the "backend" (the logical/state aspect of the app). In retrospect, this distinction might deserve its own section altogether.

dingtaoliu replied:
I think to clarify this point, maybe you can change "frontend" to "react frontend app" since the frontend is sort of its own app, separate from the Rails API server?

chrislessard replied:
Sure, sounds good to me. I'll add a commit addressing that.

Commit:
Merge branch 'master' into chrislessard-update-readme

chrislessard committed:
Clarify point on the frontend
